src/main/java/jp/ac/utokyo/rcast/karkinos/filter/SoftClipExtention.java

```diff
@@ -0,0 +1,181 @@
+limitations under the License.
+*/
+package jp.ac.utokyo.rcast.karkinos.filter;
+
+import jp.ac.utokyo.rcast.karkinos.utils.TwoBitGenomeReader;
+
+import htsjdk.samtools.Cigar;
+import htsjdk.samtools.CigarElement;
+import htsjdk.samtools.CigarOperator;
+import htsjdk.samtools.SAMRecord;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class SoftClipExtention {
+
+    //add 2020/12/17 for FFPE anneling near repeat, extends soft clipping
+    public static void extendSoftclip(SAMRecord sam, TwoBitGenomeReader tgr) {
+        try {
+            if(sam.getCigar().getCigarElements().size()>=2) {
+                leftExtention(sam, tgr);
+                rightExtention(sam, tgr);
+            }
+
+        }catch(Exception ex){
+            //could not extend soft clips
+            ex.printStackTrace();
+        }
+    }
+
+    private static void leftExtention(SAMRecord sam, TwoBitGenomeReader tgr) {
+        List<CigarElement> list = sam.getCigar().getCigarElements();
+        CigarElement first = list.get(0);
+        CigarElement second = list.get(1);
+        int extraCheckLen = 15;
+
+        if(isClipped(first)&&isM(second)&&second.getLength()>extraCheckLen){
+
+            int alignmentStart = sam.getAlignmentStart();
+            int miscount = 0;
+            int lastidx = 0;
+            int m = 0;
+            int clipidx = 0;
+            for(int n=alignmentStart;n<=alignmentStart+extraCheckLen;n++){
+                //
+                char refNuc = tgr._getGenomeNuc(n,true);
+                int idx = sam.getReadPositionAtReferencePosition(n)-1;
+                if(idx>=0 && idx<sam.getReadLength()){
+                    char readNuc = sam.getReadString().charAt(idx);
+                    if(!equalnuc(refNuc,readNuc)){
+                        miscount++;
+                        lastidx = idx;
+                    }
+                }
+                if(m==4){
+                    if(miscount>0){
+                        clipidx = lastidx;
+                    }
+                }else if(m==9){
+                    if(miscount>1){
+                        clipidx = lastidx;
+                    }
+                }else if(m==14){
+                    if(miscount>2){
+                        clipidx = lastidx;
+                    }
+                }
+                m++;
+            }
+
+            //
+            if(clipidx>0){
+
+                //
+                Cigar cgn = new Cigar();
+                CigarElement sc = new CigarElement(clipidx+1,CigarOperator.S);
+                cgn.add(sc);
+                CigarElement mc = new CigarElement((first.getLength()+second.getLength())-(clipidx+1),CigarOperator.M);
+                cgn.add(mc);
+                for(m=2; m<list.size(); m++){
+                    cgn.add(list.get(m));
+                }
+                sam.setCigar(cgn);
+                int alstart = sam.getStart()+clipidx+1;
+                sam.setAlignmentStart(alstart);
```

Comment:
`setCigar` and `setAlignmentStart` does not update `TLEN` and the mate's `PNEXT` and `MC:Z:`.
Such contradictory paired-end alignments may cause problems in the latter processes.

Reply:
Not in this case.

Reply:
Thank you for your clarification. Then there is no problem.